common/libs/keystore.js

```diff
@@ -5,10 +5,97 @@ import {
   pbkdf2Sync,
   createDecipheriv
 } from 'crypto';
-import { sha3 } from 'ethereumjs-util';
+import { decipherBuffer, decodeCryptojsSalt, evp_kdf } from './decrypt';
+import { sha3, privateToAddress } from 'ethereumjs-util';
 import scrypt from 'scryptsy';
 import uuid from 'uuid';
 
+//adapted from https://github.com/kvhnuke/etherwallet/blob/de536ffebb4f2d1af892a32697e89d1a0d906b01/app/scripts/myetherwallet.js#L342
+export function determineKeystoreType(file: string): string {
+  const parsed = JSON.parse(file);
+
+  if (parsed.encseed) return 'presale';
+  else if (parsed.Crypto || parsed.crypto) return 'v2-v3-utc';
+  else if (parsed.hash && parsed.locked === true) return 'v1-encrypted';
+  else if (parsed.hash && parsed.locked === false) return 'v1-unencrypted';
+  else if (parsed.publisher === 'MyEtherWallet') return 'v2-unencrypted';
+  else throw new Error('Invalid keystore');
+}
+
+export function isKeystorePassRequired(file: string): boolean {
+  switch (determineKeystoreType(file)) {
+    case 'presale':
+      return true;
+    case 'v1-unencrypted':
+      return false;
+    case 'v1-encrypted':
+      return true;
+    case 'v2-unencrypted':
+      return false;
+    case 'v2-v3-utc':
+      return true;
+    default:
+      return false;
+  }
+}
+
+//adapted from https://github.com/kvhnuke/etherwallet/blob/de536ffebb4f2d1af892a32697e89d1a0d906b01/app/scripts/myetherwallet.js#L218
+export function decryptPresaleToPrivKey(
+  file: string,
+  password: string
+): Buffer {
+  let json = JSON.parse(file);
+  let encseed = new Buffer(json.encseed, 'hex');
+  let derivedKey = pbkdf2Sync(
```

Comment:
Sounds like we might have to figure out a different the wallet class structure. As described in #103, this PR has `super(decryptUtcKeystoreToPkey(keystore, password));` for UTC keystores. I don't think we can have async calls in class constructors?